nucypher/policy/models.py

```diff
@@ -534,140 +534,166 @@ def __len__(self):
-                   receipt_bytes, signature)
+        return cls(bob=bob,
+                   arrangement_id=arrangement_id,
+                   tasks=tasks,
+                   alice_address=alice_address,
+                   blockhash=blockhash,
+                   receipt_signature=signature)
 
     def payload(self):
-        capsules_as_bytes = [bytes(p) for p in self.capsules]
-        capsule_signatures_as_bytes = [bytes(s) for s in self.capsule_signatures]
-        packed_receipt_and_capsules = msgpack.dumps(
-            (self.receipt_bytes,
-             msgpack.dumps(capsules_as_bytes),
-             msgpack.dumps(capsule_signatures_as_bytes),
-             self.alice_address,
-             self.alice_address_signature,
-             )
-        )
-        return bytes(self.receipt_signature) + self.bob.stamp + packed_receipt_and_capsules
+        tasks_bytes = [bytes(item) for item in self.tasks]
+        payload_elements = msgpack.dumps((tasks_bytes, self.blockhash))
+        return bytes(self.receipt_signature) + self.bob.stamp + payload_elements
 
     def complete(self, cfrags_and_signatures):
         good_cfrags = []
         if not len(self) == len(cfrags_and_signatures):
             raise ValueError("Ursula gave back the wrong number of cfrags.  "
                              "She's up to something.")
 
-        alice_address_signature = bytes(self.alice_address_signature)
         ursula_verifying_key = self.ursula.stamp.as_umbral_pubkey()
 
-        for counter, capsule in enumerate(self.capsules):
-            cfrag, signature = cfrags_and_signatures[counter]
-
-            # Validate CFrag metadata
-            capsule_signature = bytes(self.capsule_signatures[counter])
-            metadata_input = capsule_signature + alice_address_signature
+        for task, (cfrag, reencryption_signature) in zip(self.tasks, cfrags_and_signatures):
+            # Validate re-encryption metadata
+            metadata_input = bytes(task.signature)
             metadata_as_signature = Signature.from_bytes(cfrag.proof.metadata)
             if not metadata_as_signature.verify(metadata_input, ursula_verifying_key):
-                raise InvalidSignature("Invalid metadata for {}.".format(cfrag))
+                raise InvalidSignature(f"Invalid metadata for {cfrag}.")
+                # TODO: Instead of raising, we should do something
 
-            # Validate work order response signatures
-            if signature.verify(bytes(cfrag) + bytes(capsule), ursula_verifying_key):
+            # Validate re-encryption signatures
+            if reencryption_signature.verify(bytes(cfrag), ursula_verifying_key):
                 good_cfrags.append(cfrag)
             else:
-                raise InvalidSignature("{} is not properly signed by Ursula.".format(cfrag))
-        else:
-            self.completed = maya.now()
-            return good_cfrags
+                raise InvalidSignature(f"{cfrag} is not properly signed by Ursula.")
```

Comment:
I'm not exactly sure where this PR falls in the timeline of slashing integration, but I noticed that there is little test coverage on these `raise` conditions, including this one.

Reply:
#626